[wasm][debugger] - RuntimeError: memory access out of bounds

pavelsavara opened the pull request:
Fixes https://github.com/dotnet/runtime/issues/51276

- fix null reference in describe_value() + add test for it
- fix null reference in PDB load
- improved handling of targetCrashed in Inspector

On top of that
- improve Makefile to allow passing test arguments together with filters
- add windows chrome to probe path
- include debugging payload into test failure exception for easier debugging
- make it possible to compile with WasmBuildNative=true
- make it possible to compile with Debug runtime
- enable Mono logging in debug mode

pavelsavara commented:
There are few more places which throw memory access out of bounds, converted to draft.

pavelsavara commented:
/azp help

pavelsavara commented:
/azp list

pavelsavara commented:
/azp run runtime

CoffeeFlux commented:
/azp run runtime

radical commented on src/mono/wasm/debugger/DebuggerTestSuite/Inspector.cs:
```diff
@@ -158,14 +159,26 @@ async Task OnMessage(string method, JObject args, CancellationToken token)
                 case "Runtime.consoleAPICalled":
                     _logger.LogInformation(FormatConsoleAPICalled(args));
                     break;
+                case "Inspector.targetCrashed":
+                case "Runtime.exceptionThrown":
+                    _logger.LogWarning($"Failing all waiters because: {method}: {args}");
+                    fail = true;
+                    break;
```
A couple more could be added for this:

```csharp
		static string[] FailOnMethods = new string[] {
			"Runtime.exceptionThrown",
			"Inspector.targetCrashed",
			"Inspector.targetReloadedAfterCrash",
			"Inspector.detached"
		};
```

radical commented on src/mono/wasm/debugger/DebuggerTestSuite/Inspector.cs:
```diff
@@ -158,14 +159,26 @@ async Task OnMessage(string method, JObject args, CancellationToken token)
+                    _logger.LogWarning($"Failing all waiters because: {method}: {args}");
+                    fail = true;
+                    break;
+                case "Debugger.scriptParsed":
+                case "Debugger.resumed":
+                case "Debugger.breakpointResolved":
+                case "Runtime.executionContextCreated":
+                    break;
+                default:
+                    _logger.LogWarning(method);
+                    break;
```
Why are these needed?

pavelsavara replied:
I wanted to see any unknow in the log as warning. So, those which are known to be OK, should not spam the log.

radical replied:
This doesn't expect to handle every message, so this doesn't need to be a warning. For example, you could have messages that are being waited upon via `On(..)`, which gets handled at line 178, and those should not generate a warning.

pavelsavara replied:
improved

radical commented on src/mono/wasm/debugger/DebuggerTestSuite/SteppingTests.cs:
```diff
@@ -250,6 +250,21 @@ public async Task InspectLocalsInPreviousFramesDuringSteppingIn(bool use_cfo)
         }
 
+        [Fact]
+        public async Task InspectNullLocalsDuringSteppingIn()
```
Which null local is being tested here?

pavelsavara replied:
I renamed it and added another explicit test for the other null

radical commented on src/mono/wasm/debugger/tests/debugger-test/debugger-test.csproj:
```diff
@@ -7,6 +7,10 @@
     <WasmBuildAppDependsOn>PrepareForWasmBuildApp;$(WasmBuildAppDependsOn)</WasmBuildAppDependsOn>
   </PropertyGroup>
 
+  <PropertyGroup Condition="'$(RuntimeConfiguration)'=='Debug'">
```
The property changes in this file would be useful for other test assemblies too. They can be moved to `tests/Directory.Build.props`.

radical commented on src/mono/wasm/debugger/DebuggerTestSuite/DebuggerTestBase.cs:
```diff
@@ -81,6 +82,10 @@ public DebuggerTestBase(string driver = "debugger-driver.html")
 
             startTask = TestHarnessProxy.Start(FindChromePath(), DebuggerTestAppPath, driver);
+
+            // the debugger is working in locale of the debugged application. For example Datetime.ToString()
+            // we want the test to mach it. We are also starting chrome with --lang=en-US
+            System.Globalization.CultureInfo.CurrentCulture = new System.Globalization.CultureInfo("en-US");
```
IMHO, this, and the corresponding `Makefile` change should be in a separate PR.

radical commented on src/mono/mono/mini/mini-wasm-debugger.c:
```diff
@@ -1070,6 +1074,13 @@ describe_value(MonoType * type, gpointer addr, int gpflags)
 		case MONO_TYPE_ARRAY:
 		case MONO_TYPE_CLASS: {
 			MonoObject *obj = *(MonoObject**)addr;
+			if (!obj) {
```
We need explicit tests for both the cases being fixed.

radical commented on src/mono/wasm/debugger/DebuggerTestSuite/Inspector.cs:
```diff
@@ -158,14 +159,28 @@ async Task OnMessage(string method, JObject args, CancellationToken token)
+                case "Inspector.targetReloadedAfterCrash":
+                case "Inspector.detached":
+                case "Runtime.exceptionThrown":
+                    _logger.LogWarning($"Failing all waiters because: {method}: {args}");
```
These can be expected (eg. when testing for these messages), so maybe we can reduce this to `LogInformation`? Also, this should be getting logged already.

pavelsavara replied:
You are right that Runtime.exceptionThrown is possibly expected. I returned original Debug log level.
The browser crash is not expected, but it actually happens. That should be warning.

radical replied:
It should fail the test completely, so we shouldn't need the warning.

pavelsavara replied:
improved

radical commented on src/mono/wasm/debugger/DebuggerTestSuite/Inspector.cs:
```diff
@@ -158,14 +159,22 @@ async Task OnMessage(string method, JObject args, CancellationToken token)
+            else if (fail)
             {
-                _logger.LogDebug($"Failing all waiters because: {method}: {args}");
                 FailAllWaiters(new ArgumentException(args.ToString()));
```
We could instead pass `new ArgumentException ($"{method}: {args}")` to `FailAllWaiters` here. It will always get surfaced then, and we don't need to do the `LogDebug` either.

And we want to do that for all 4 cases where it's crashing, or exception is thrown etc.

pavelsavara replied:
There are places which parse back the Json. See https://github.com/pavelsavara/runtime/blob/pr_wasm_debugger_crash/src/mono/wasm/debugger/DebuggerTestSuite/Tests.cs#L57

radical commented on src/mono/wasm/debugger/DebuggerTestSuite/SteppingTests.cs:
```diff
@@ -250,6 +250,38 @@ public async Task InspectLocalsInPreviousFramesDuringSteppingIn(bool use_cfo)
+        }
+
+        [Fact]
+        public async Task InspectNullBoxedLocalsDuringSteppingIn()
```
Do we know why this isn't getting caught in `InspectBoxedLocals`, which has a `o_null` case in  https://github.com/pavelsavara/runtime/blob/pr_wasm_debugger_crash/src/mono/wasm/debugger/tests/debugger-test/debugger-test.cs#L358 ?

radical replied:
Ugh.. my fault

pavelsavara replied:
@radical Could you please create separate issue about it? 
I'm not comfortable mixing it with this change. 
Especially because I don't know what is the "correct" shape of the json for null object.

radical commented on src/mono/wasm/debugger/DebuggerTestSuite/SteppingTests.cs:
```diff
@@ -250,6 +250,38 @@ public async Task InspectLocalsInPreviousFramesDuringSteppingIn(bool use_cfo)
         }
 
+        [Fact]
+        public async Task InspectNullStringBuilderLocalsDuringSteppingIn()
```
What is special about `StringBuilder`? We have other tests for null object locals, IIRC.

pavelsavara replied:
debugger handles `object` as `MONO_TYPE_OBJECT` and `StringBuilder` as `MONO_TYPE_CLASS`. They are 2 different code paths. We fixed both, we are testing both.

radical commented:
IIUC, the real issue here is when you are paused "before the method has started" .. and try to look at a local, then that is NULL.

I would suggest:

- Test something like:
```csharp
             await SetBreakpointInMethod("debugger-test.dll", "WithStringBuilder", "OuterMethod", 0);
             await EvaluateAndCheck("window.setTimeout(function() { invoke_static_method('[debugger-test] WithStringBuilder:OuterMethod'); }, 1);", "dotnet://debugger-test.dll/debugger-test.cs", 561, 5, "OuterMethod");

             await StepAndCheck(StepKind.Into, "dotnet://debugger-test.dll/debugger-test.cs", 566, 5, "InnerMethod", times: 2,
                 locals_fn: async (locals) => await CheckProps(locals, new
                 {
                     sb = TObject("System.Text.StringBuilder", is_null: true)
                 }, "locals"));

             await StepAndCheck(StepKind.Into, "dotnet://debugger-test.dll/debugger-test.cs", 567, 9, "InnerMethod", times: 1,
                 locals_fn: async (locals) => await CheckProps(locals, new
                 {
                     sb = TObject("System.Text.StringBuilder")
                 }, "locals"));
```

  - this breaks at `OuterMethod`
  - steps into `InnerMethod`, and pauses at the opening brace
  - We check for locals there, which should be null (this should cause the crash without your fix)
  - Then we step once again, so `sb` gets initialized
  - And we can check that it's not null now

- I would also suggest using a simple type, like one of the classes in the debugger tests, instead of `StringBuilder`
- And add similar tests for:
  - string, primitive types
  - valuetypes
  - generic types

pavelsavara committed:
- fix null reference in describe_value() + add test for it
- fix null reference in PDB load
- improved handling of targetCrashed in Inspector
- improve Makefile to allow passing test arguments together with filters
- add windows chrome to probe path
- include debugging payload into test failure exception for easier debugging
- make it possible to compile with WasmBuildNative=true
- make it possible to compile with Debug runtime
- enable Mono logging in debug mode

lewing reviewed (commented):
Lots of nice improvements here.  I'm doing a local testing before approving because of the CI situation.

radical commented:
For the cases that are actually being fixed here, IMHO, the new tests should be explicit that they are testing the case of looking at locals before they get initialized. So, maybe name it like that, and the test code can be like:

```csharp
1        public static int OuterMethod()
2        {
3			 MONO_TYPE_CLASS mtc;
4			 Console.WriteLine("break here"); // breakpoint here             
5            mtc = new MONO_TYPE_CLASS();
6            Console.WriteLine($"mtc: {mtc}");
7        }
```

Set breakpoint on line 4. Get the locals there:

```csharp
var pause_location = await EvaluateAndCheck(...
var locals = await GetProperties(pause_location["callFrames"][0]["callFrameId"].Value<string>());
```

Otherwise, the tests seem like they are accidentally checking the case.

radical commented:
With a clean DEBUG runtime build, the `DebuggerTests.AssignmentTests.InspectDecimalAssignmentsDuringSteppingIn` fails with:

```
info: DebuggerTests.Inspector[0] console.trace: * Assertion at /Users/radical/dev/r2/src/mono/mono/mini/debugger-engine.c:765, condition `req->refcount' not met
info: DebuggerTests.Inspector[0] console.error: * Assertion at /Users/radical/dev/r2/src/mono/mono/mini/debugger-engine.c:765, condition `req->refcount' not met
```

```json
{
  "timestamp": 1620228774790.821,
  "exceptionDetails": {
    "exceptionId": 3,
    "text": "Uncaught",
    "lineNumber": 37,
    "columnNumber": 2,
    "scriptId": "8",
    "url": "http://localhost:9400/dotnet.js",
    "stackTrace": {
      "callFrames": [
        {
          "functionName": "quit_",
          "scriptId": "8",
          "url": "http://localhost:9400/dotnet.js",
          "lineNumber": 37,
          "columnNumber": 2
        },
        {
          "functionName": "exit",
          "scriptId": "8",
          "url": "http://localhost:9400/dotnet.js",
          "lineNumber": 11616,
          "columnNumber": 2
        },
        {
          "functionName": "_exit",
          "scriptId": "8",
          "url": "http://localhost:9400/dotnet.js",
          "lineNumber": 6056,
          "columnNumber": 6
        },
        {
          "functionName": "monoeg_assertion_message",
          "scriptId": "11",
          "url": "http://localhost:9400/dotnet.wasm",
          "lineNumber": 0,
          "columnNumber": 1889128
        },
        {
          "functionName": "mono_assertion_message",
          "scriptId": "11",
          "url": "http://localhost:9400/dotnet.wasm",
          "lineNumber": 0,
          "columnNumber": 1889214
        },
        {
          "functionName": "mono_de_ss_req_release",
          "scriptId": "11",
          "url": "http://localhost:9400/dotnet.wasm",
          "lineNumber": 0,
          "columnNumber": 1830351
        },
        {
          "functionName": "mono_de_process_single_step",
          "scriptId": "11",
          "url": "http://localhost:9400/dotnet.wasm",
          "lineNumber": 0,
          "columnNumber": 1831800
        },
        {
          "functionName": "mono_wasm_single_step_hit",
          "scriptId": "11",
          "url": "http://localhost:9400/dotnet.wasm",
          "lineNumber": 0,
          "columnNumber": 1810692
        },
        {
          "functionName": "do_debugger_tramp",
          "scriptId": "11",
          "url": "http://localhost:9400/dotnet.wasm",
          "lineNumber": 0,
          "columnNumber": 171020
        },
        {
          "functionName": "interp_exec_method",
          "scriptId": "11",
          "url": "http://localhost:9400/dotnet.wasm",
          "lineNumber": 0,
          "columnNumber": 145696
        },
        {
          "functionName": "interp_runtime_invoke",
          "scriptId": "11",
          "url": "http://localhost:9400/dotnet.wasm",
          "lineNumber": 0,
          "columnNumber": 36309
        },
        {
          "functionName": "mono_jit_runtime_invoke",
          "scriptId": "11",
          "url": "http://localhost:9400/dotnet.wasm",
          "lineNumber": 0,
          "columnNumber": 1562745
        },
        {
          "functionName": "do_runtime_invoke",
          "scriptId": "11",
          "url": "http://localhost:9400/dotnet.wasm",
          "lineNumber": 0,
          "columnNumber": 902006
        },
        {
          "functionName": "mono_runtime_try_invoke",
          "scriptId": "11",
          "url": "http://localhost:9400/dotnet.wasm",
          "lineNumber": 0,
          "columnNumber": 904955
        },
        {
          "functionName": "mono_runtime_invoke",
          "scriptId": "11",
          "url": "http://localhost:9400/dotnet.wasm",
          "lineNumber": 0,
          "columnNumber": 921755
        },
        {
          "functionName": "mono_wasm_invoke_method",
          "scriptId": "11",
          "url": "http://localhost:9400/dotnet.wasm",
          "lineNumber": 0,
          "columnNumber": 2954696
        },
        {
          "functionName": "",
          "scriptId": "8",
          "url": "http://localhost:9400/dotnet.js",
          "lineNumber": 1430,
          "columnNumber": 21
        },
        {
          "functionName": "ccall",
          "scriptId": "8",
          "url": "http://localhost:9400/dotnet.js",
          "lineNumber": 644,
          "columnNumber": 17
        },
        {
          "functionName": "",
          "scriptId": "8",
          "url": "http://localhost:9400/dotnet.js",
          "lineNumber": 656,
          "columnNumber": 11
        },
        {
          "functionName": "managed__debugger_test__DebuggerTests_MONO_TYPE_VALUETYPE2_OuterMethod",
          "scriptId": "48",
          "url": "https://mono-wasm.invalid/managed__debugger_test__DebuggerTests_MONO_TYPE_VALUETYPE2_OuterMethod",
          "lineNumber": 14,
          "columnNumber": 39
        },
        {
          "functionName": "invoke_static_method",
          "scriptId": "5",
          "url": "http://localhost:9400/debugger-driver.html",
          "lineNumber": 25,
          "columnNumber": 10
        },
        {
          "functionName": "",
          "scriptId": "47",
          "url": "",
          "lineNumber": 0,
          "columnNumber": 31
        }
      ]
    },
    "exception": {
      "type": "object",
      "className": "ExitStatus",
      "description": "ExitStatus",
      "objectId": "7004380125509995368.2.141",
      "preview": {
        "type": "object",
        "description": "ExitStatus",
        "overflow": false,
        "properties": [
          {
            "name": "name",
            "type": "string",
            "value": "ExitStatus"
          },
          {
            "name": "message",
            "type": "string",
            "value": "Program terminated with exit(0)"
          },
          {
            "name": "status",
            "type": "number",
            "value": "0"
          }
        ]
      }
    },
    "executionContextId": 2
  }
}
```

pavelsavara committed:
code review feedback

pavelsavara commented:
We agreed with @radical that he would open another issue for 'req->refcount not met' problem as it reproduces even without the runtime changes on this PR.

I renamed files and methods to make it clear what is just preparation and what is actually being tested. I also added comments.

radical commented:
> I renamed files and methods to make it clear what is just preparation and what is actually being tested. I also added comments.

Sorry, I wasn't clear. I meant that these tests are not related to Stepping, but are about trying to read values for locals before they get initialized. And so, I was suggesting that we should rename the *xunit* test names to reflect that.

radical commented on src/mono/wasm/debugger/DebuggerTestSuite/AssignmentTests.cs:
```diff
@@ -0,0 +1,238 @@
+        async Task InspectBoolAssignmentsDuringSteppingIn()
+        {
+            await InspectAssignmentDuringSteppingIn("MONO_TYPE_BOOLEAN",
+                (locals) => CheckBool(locals, "r", default),
```
.. and this can use `TBool` then

radical commented on src/mono/wasm/debugger/DebuggerTestSuite/AssignmentTests.cs:
```diff
@@ -0,0 +1,238 @@
+        async Task InspectCharAssignmentsDuringSteppingIn()
+        {
+            await InspectAssignmentDuringSteppingIn("MONO_TYPE_CHAR",
+                (locals) => CheckSymbol(locals, "r", "0 '\u0000'"),
```
`TSymbol("0 '\u0000'"), TSymbol("97 'a'")`

radical commented on src/mono/wasm/debugger/DebuggerTestSuite/AssignmentTests.cs:
```diff
@@ -0,0 +1,238 @@
+        async Task InspectFunctionAssignmentsDuringSteppingIn()
+        {
+            await InspectAssignmentDuringSteppingIn("MONO_TYPE_GENERICINST",
+                (locals) => CheckObject(locals, "r", "System.Func<int>", is_null: true),
+                (locals) => CheckObject(locals, "r", "System.Func<int>", description: "int Prepare ()")
```
For this:
```csharp
 "MONO_TYPE_GENERICINST", TObject("System.Func<int>", is_null: true), TDelegate("System.Func<int>", "int Prepare ()")
```

radical commented on src/mono/wasm/debugger/DebuggerTestSuite/AssignmentTests.cs:
```diff
@@ -0,0 +1,238 @@
+        async Task InspectFunctionPtrAssignmentsDuringSteppingIn()
+        {
+            await InspectAssignmentDuringSteppingIn("MONO_TYPE_FNPTR",
+                (locals) => CheckSymbol(locals, "r", "(*()) 0"),
+                (locals) => { } // CheckObject(locals, "r", "(*()) 0x22dd230") // FixMe: Could we do better ?
```
```csharp
{ "MONO_TYPE_FNPTR", TPointer("*()",  is_null: true), TPointer("*()") },
```

radical commented on src/mono/wasm/debugger/DebuggerTestSuite/AssignmentTests.cs:
```diff
@@ -0,0 +1,238 @@
+        async Task InspectIntPtrAssignmentsDuringSteppingIn()
+        {
+            await InspectAssignmentDuringSteppingIn("MONO_TYPE_PTR",
+                (locals) => CheckSymbol(locals, "r", "(int*) 0"),
+                (locals) => { } //CheckObject(locals, "r", "(int*) 0x211c058") // FixMe: Could we do better ?
```
```csharp
 "MONO_TYPE_PTR",   TPointer("int*", is_null: true), TPointer("int*") }
```

radical commented on src/mono/wasm/debugger/DebuggerTestSuite/AssignmentTests.cs:
```diff
@@ -0,0 +1,238 @@
+        async Task InspectStructAssignmentsDuringSteppingIn()
+        {
+            await InspectAssignmentDuringSteppingIn("MONO_TYPE_VALUETYPE",
+                (locals) => CheckValueType(locals, "r", "DebuggerTests.Point"),
+                (locals) => CheckValueType(locals, "r", "DebuggerTests.Point")
```
`TValueType("DebuggerTests.Point")`

radical commented on src/mono/wasm/debugger/DebuggerTestSuite/AssignmentTests.cs:
```diff
@@ -0,0 +1,238 @@
+        async Task InspectDecimalAssignmentsDuringSteppingIn()
+        {
+            await InspectAssignmentDuringSteppingIn("MONO_TYPE_VALUETYPE2",
+                (locals) => CheckValueType(locals, "r", "System.Decimal", description: "0"),
+                (locals) => CheckValueType(locals, "r", "System.Decimal", description: "1.1")
```
`TValueType("System.Decimal", "0")`

radical commented on src/mono/wasm/debugger/DebuggerTestSuite/AssignmentTests.cs:
```diff
@@ -0,0 +1,238 @@
+
+        private async Task InspectAssignmentDuringSteppingIn(string clazz, Action<JToken> checkDefault, Action<JToken> checkValue)
+        {
+            await SetBreakpointInMethod("debugger-test", "DebuggerTests." + clazz, "Prepare", 2);
```
nit: Breakpoint can be set directly in the `TestedMethod`, `offset=0`, thus avoiding potential influence of *stepping* into `TestedMethod`. (maybe a bit paranoid ;)

pavelsavara replied:
Some of the tests have custom TestedMethod for them, which would require custom breakpoint location and complicate stuff. I will keep it as is.

radical commented on src/mono/wasm/debugger/DebuggerTestSuite/AssignmentTests.cs:
```diff
@@ -0,0 +1,238 @@
+            await StepAndCheck(StepKind.Into, "dotnet://debugger-test.dll/debugger-assignment-test.cs", -1, -1, "TestedMethod",
+                locals_fn: (locals) =>
+                {
+                    Console.WriteLine(locals);
```
These can be dropped.

pavelsavara replied:
done

radical commented on src/mono/wasm/debugger/tests/Directory.Build.props:
```diff
@@ -16,4 +16,9 @@
     <MicrosoftNetCoreAppRuntimePackRidDir>$(ArtifactsBinDir)microsoft.netcore.app.runtime.browser-wasm\$(RuntimeConfiguration)\runtimes\browser-wasm\</MicrosoftNetCoreAppRuntimePackRidDir>
   </PropertyGroup>
+  
+  <PropertyGroup Condition="'$(RuntimeConfiguration)'=='Debug'">
+    <WasmDebugLevel>5</WasmDebugLevel>
+  </PropertyGroup>
```
Um this seems excessive to have by default. It makes the test output large, and difficult to read.

IMHO, whoever is running the tests can set it from the command line, eg. `make run-debugger-tests MSBUILD_ARGS="/p:WasmDebugLevel=5"`.

pavelsavara replied:
removed and added into readme

radical reviewed (changes requested):
Other than the comments, looks good. Thank you for testing on windows, and with `Debug`!

radical reviewed (approved):
Other than the cleanup comments, looks good 👍

pavelsavara committed:
code review

pavelsavara commented on src/mono/wasm/debugger/DebuggerTestSuite/AssignmentTests.cs:
```diff
@@ -0,0 +1,238 @@
+{
+    public class AssignmentTests : DebuggerTestBase
+    {
+        [Fact]
```
Thank you 👍

pavelsavara committed:
lint

radical commented on src/mono/wasm/debugger/DebuggerTestSuite/AssignmentTests.cs:
```diff
@@ -0,0 +1,66 @@
+                locals_fn: (locals) =>
+                {
+                    Assert.Equal(2, locals.Count());
+                    Check(locals, "r", checkDefault);
```
Instead of the new overloads that use `CheckValue`, you can use:
```csharp
await CheckProps(locals, new
{
  r = checkDefault
}, "default", num_fields: 2);
```

But since this needs to be async, you will need to do:
```csharp
var pause_location = await StepAndCheck(StepKind.Into, "dotnet://debugger-test.dll/debugger-assignment-test.cs", -1, -1, "TestedMethod");
var locals = await GetProperties(pause_location["callFrames"][0]["callFrameId"].Value<string>());

await CheckProps(locals, new
{
  r = checkDefault
}, "default", num_fields: 2);
```

lewing reviewed (approved):
Looks good, we can do any more clean-up in followups
